runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -641,8 +641,10 @@ protected final void setBundleFinishedCallback(Runnable callback) {
   @Override
   public final void processElement(StreamRecord<WindowedValue<InputT>> streamRecord) {
     checkInvokeStartBundle();
+    long oldHold = keyCoder != null ? keyedStateInternals.minWatermarkHoldMs() : -1L;
     doFnRunner.processElement(streamRecord.getValue());
     checkInvokeFinishBundleByCount();
+    emitWatermarkIfHoldChanged(oldHold);
```

Comment:
Is this required on every element? I'd rather trigger this only if we set / remove a hold.

Reply:
This is very cheap call and actually tests exactly if hold was set (actually reset).